curiosrc/market/lmrpc/lmrpc.go

```diff
@@ -0,0 +1,422 @@
+package lmrpc
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/jackc/pgx/v5"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/types"
+	cumarket "github.com/filecoin-project/lotus/curiosrc/market"
+	"github.com/filecoin-project/lotus/curiosrc/market/fakelm"
+	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+	"github.com/filecoin-project/lotus/lib/nullreader"
+	"github.com/filecoin-project/lotus/metrics/proxy"
+	"github.com/filecoin-project/lotus/node"
+	"github.com/filecoin-project/lotus/node/config"
+	"github.com/filecoin-project/lotus/storage/paths"
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+var log = logging.Logger("lmrpc")
+
+func ServeCurioMarketRPCFromConfig(db *harmonydb.DB, full api.FullNode, cfg *config.CurioConfig) error {
+	return forEachMarketRPC(cfg, func(maddr string, listen string) error {
+		addr, err := address.NewFromString(maddr)
+		if err != nil {
+			return xerrors.Errorf("parsing actor address: %w", err)
+		}
+
+		go func() {
+			err := ServeCurioMarketRPC(db, full, addr, cfg, listen)
+			if err != nil {
+				log.Errorf("failed to serve market rpc: %s", err)
+			}
+		}()
+
+		return nil
+	})
+}
+
+func MakeTokens(cfg *config.CurioConfig) (map[address.Address]string, error) {
+	out := map[address.Address]string{}
+
+	err := forEachMarketRPC(cfg, func(smaddr string, listen string) error {
+		ctx := context.Background()
+
+		laddr, err := net.ResolveTCPAddr("tcp", listen)
+		if err != nil {
+			return xerrors.Errorf("net resolve: %w", err)
+		}
+
+		if len(laddr.IP) == 0 {
+			// set localhost
+			laddr.IP = net.IPv4(127, 0, 0, 1)
+		}
+
+		// need minimal provider with just the config
+		lp := fakelm.NewLMRPCProvider(nil, nil, address.Undef, 0, 0, nil, nil, cfg)
+
+		tok, err := lp.AuthNew(ctx, api.AllPermissions)
+		if err != nil {
+			return err
+		}
+
+		// parse listen into multiaddr
+		ma, err := manet.FromNetAddr(laddr)
+		if err != nil {
+			return xerrors.Errorf("net from addr (%v): %w", laddr, err)
+		}
+
+		maddr, err := address.NewFromString(smaddr)
+		if err != nil {
+			return xerrors.Errorf("parsing actor address: %w", err)
+		}
+
+		token := fmt.Sprintf("%s:%s", tok, ma)
+		out[maddr] = token
+
+		return nil
+	})
+
+	return out, err
+}
+
+func forEachMarketRPC(cfg *config.CurioConfig, cb func(string, string) error) error {
+	for n, server := range cfg.Subsystems.MarketRPCServers {
```

Comment:
Could we rename this to cfg.Subsystems.BoostAdapter ?

`MarketRPCServers` will be confused with Curio's (future) Market node's config. These two will live side-by-side for some time.

Reply:
Yeah, I was a bit torn on that one as technically this supports more than just boost (and I'm vaguely aware that there are non-boost market nodes), but yeah for clarity it may make sense to just call this BoostAdapter and just say in the comment that it's boost-style rpc.